pep-9999.rst

```diff
@@ -0,0 +1,988 @@
+     changes to ``sysconfig`` won't be visible, and there
+     will only be one shared scheme.
+
+Specification
+=============
+
+Marking an interpreter as using an external package manager
+-----------------------------------------------------------
+
+Before a Python package installer (that is, a tool such as
+pip - not an external tool such as apt) installs a package
+into a certain Python context, it should make the following
+checks by default:
+
+1. Is it running outside of a virtual environment? It can
+   determine this by whether ``sys.prefix ==
+   sys.base_prefix`` (but see `Backwards Compatibility`_).
+
+2. Is there a ``EXTERNALLY-MANAGED`` file in the directory
+   identified by ``sysconfig.get_path("stdlib",
+   sysconfig.get_default_scheme())``
+
+If both of these conditions are true, the installer should
+exit with an error message indicating that package
+installation into this Python interpreter's directory are
+disabled outside of a virtual environment.
+
+The installer should have a way for the user to override
+these rules, such as a command-line flag
+``--break-system-packages``.
+
+If the contents of the ``EXTERNALLY-MANAGED`` file parses as
+an ``email.message.EmailMessage`` using the routines in
+``email.parser``, the installer should output an error
+message from that file as part of its error. If
+``locale.getlocale(locale.LC_MESSAGES)`` returns
+non-``None`` and the first element is a string of the form
+``xx_YY``, and the file contains a header variable
+``Error-xx_YY`` or failing that ``Error-xx``, then the
+installer should use the value of that header as the error.
+Otherwise, it should use the body of the message as an
+error.
+
+If the file does not parse as an ``EmailMessage``, then the
+installer should ignore the parse failure and instead just
+use a pre-defined error message of its own, which should
+suggest that the user create a virtual environment to
+install packages.
+
+Software distributors who have a non-Python-specific package
+manager that manages libraries in the ``sys.path`` of their
+Python package should, in general, ship a
+``EXTERNALLY-MANAGED`` file in their standard library
+directory. For instance, Debian may ship a file in
+``/usr/lib/python3.9/EXTERNALLY-MANAGED`` consisting of
+something like
+
+::
+
+    To install Python packages systemwide, try apt install
+    python3-xyz, where xyz is the package you are trying to
+    install.
+
+    If you wish to install a non-Debian-packaged Python
+    package, create a virtual environment using python3 -m
+    venv path/to/venv. Then use path/to/venv/bin/python and
+    path/to/venv/bin/pip. Make sure you have python3-full
+    installed.
+
+    If you wish to install a non-Debian packaged Python
+    application, it may be easiest to use pipx install xyz,
+    which will manage a virtual environment for you. Make
+    sure you have pipx installed.
+
+    See /usr/share/doc/python3.9/README.venv for more
+    information.
```

Comment:
This does not parse as an `EmailMessage`, but as a `Message`. By the rule above rule, this text would then be ignored.

It would probably be good to use an example with headers to show how paragraphs are handled in headers.

Reply:
FWIW, the format is gonna be to match METADATA files.

That's a bit underspecified (see start of https://packaging.python.org/specifications/core-metadata/). Basically everything uses the exact same code for this, which looks like: https://installer.readthedocs.io/en/latest/api/utils/#installer.utils.parse_metadata_file

Reply:
I'm changing this to link to https://packaging.python.org/specifications/core-metadata/ (as PEP 566 does), and mentioning that this happens to be defined as stdlib `email.parser` with `policy=compat32`. I'm worried that an example with headers will be too long here and disrupt the flow of reading, and I think deferring to the core metadata spec etc. will explain how to implement headers with paragraphs.

I'm also a bit confused in that https://packaging.python.org/specifications/core-metadata/#description says that you can use seven spaces and a pipe character to continue long paragraphs, but,
```python
>>> m = installer.utils.parse_metadata_file("""Description: This project provides powerful math functions
...         |For example, you can use `sum()` to sum numbers:
...         |
...         |Example::
...         |
...         |    >>> sum(1, 2)
...         |    3
...         |""")
>>> m["Description"]
'This project provides powerful math functions\n        |For example, you can use `sum()` to sum numbers:\n        |\n        |Example::\n        |\n        |    >>> sum(1, 2)\n        |    3\n        |'
```

Is this behavior specific to the interpretation of the Description field and not part of the document format in general?

Reply:
The long string logic of the metadata format is *extremely* underspecified and buggy, which is one of the biggest reasons I strongly dislike using that format elsewhere. Almost no metadata files in the real world actually uses that, and when they do it's an exercise of trail and error figuring out what is actually being converted to what.

Reply:
OK, that's as good a reason as any to decide against that format, since `configparser` is [very clear about multi-line handling](https://docs.python.org/3/library/configparser.html#supported-ini-file-structure):

```python
>>> import configparser
>>> c = configparser.ConfigParser()
>>> c.read_string("[externally-managed]\nError: pip install is prohibited outside virtualenvs\n\n Virtualenvs are cool")
>>> c["externally-managed"]["error"]
'pip install is prohibited outside virtualenvs\n\nVirtualenvs are cool'
```

I'll make this update and I'll also try to come up with a multi-language example.